Iteration3/master/app.py

```diff
@@ -0,0 +1,155 @@
+import asyncio
+import logging
+import threading
+
+import counter
+import httpx
+from flask import Flask, jsonify, request
+from settings import slaves_ip_addresses
+
+app = Flask(__name__)
+MESSAGES = []
+MESSAGE_ID = counter.FastWriteCounter()
+MY_RETRY = 3
+
+logging.basicConfig(format="%(asctime)s - %(message)s", level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+@app.route("/", methods=["POST", "GET"])
+async def main():
+    if request.method == "POST":
+        try:
+            write_concern: int = int(request.json.get("write_concern"))
+        except (TypeError, ValueError):
+            return jsonify(
+                {"message": "`write_concern` argument should be integer"}
+            )
+        message: str = request.json.get("message")
+
+        done = asyncio.Event()
+        done.write_concern = write_concern
+
+        MESSAGES.append(message)
+
+        if write_concern == 1:
+            for ip_address in slaves_ip_addresses():
+                send_replication_request(ip_address, message, MESSAGE_ID.value)
+            MESSAGE_ID.increment()
+            return jsonify({"message": "successful"})
+
+        common_tasks: list[tuple[asyncio.Task, str]] = []
+        for ip_address in slaves_ip_addresses():
+            task = asyncio.create_task(
+                replicate_on_slaves(
+                    ip_address, message, done, MESSAGE_ID.value
+                )
+            )
+            logger.info(
+                "Task for replication on slave:%s with ID:%d has been created",
+                ip_address,
+                MESSAGE_ID.value,
+            )
+            common_tasks.append((task, ip_address))
+        MESSAGE_ID.increment()
+
+        await done.wait()
+
+        # For tasks that were not finished we create threads in which
+        # we will make new requests to slaves for replication to ensure that
+        # slave has received our message.
+        if done.write_concern > 1:
+            for task, ip_address in common_tasks:
+                if not task.done():
+                    asyncio.create_task(do_retry_request(ip_address, message))
```

Comment:
Task that you are submitting here won't be done, because main thread will die before they will have time to finish. You should consider using threads here instead as I were doing previously on the second iteration.

Reply:
I also have some issues with this part. In ide it works good but under docker sounds like you said

Reply:
resolved in c78054bc0a202ffb1e889cd723865e480abdcffc